# Example of bipartite networks using the data generated by
# Sweet potato virome in Africa. 
# --------------------------------------------------------------------------------------
setwd("/Users/ricardoi//Documents/Documents-ehecatl/git_db/BiNeVi_deprecated/iBiNeVI/")

#----- Loading libraries -------
library(bipartite)
library(igraph)
library(dplyr)
library(plyr)
library(viridis)
library(reshape2)
library(tidyr)
library(openxlsx)

# ----- Adding metadata --------
metadata <- read.xlsx("data/CIP/2_Viroma_web_180319_final.xlsx", sheet = 2)
metadata <- metadata %>% select_if(~!all(is.na(.)))
metadata

length(metadata$`Altitude.(masl)`)
hl <- metadata[which(metadata$`Altitude.(masl)` < 1000),] # hot land Hoehen index
hls<- c(table(hl$Department), Total= sum(table(hl$Department)))
cl <- metadata[which(metadata$`Altitude.(masl)` > 2000 & metadata$`Altitude.(masl)` <3400), ] # cold land Hoehen index
cls <- c(table(cl$Department), Total= sum(table(cl$Department)))
fl <- metadata[which(metadata$`Altitude.(masl)` > 3400),] # frozen land Hoehen index
fls <- c(table(fl$Department), Total= sum(table(fl$Department)))

metadata$altzones <- ifelse(metadata$`Altitude.(masl)` < 1000, "hotland",
  ifelse(metadata$`Altitude.(masl)` > 1000 & metadata$`Altitude.(masl)` <2000, "templand", 
    ifelse(metadata$`Altitude.(masl)` > 2000 & metadata$`Altitude.(masl)` <3400, "coldland",
      ifelse(metadata$`Altitude.(masl)` > 3400 & metadata$`Altitude.(masl)` <5000, "frozenland",
        ifelse(metadata$`Altitude.(masl)` > 5000, "snowline", "out of range")))))

#### Summary statistics
#   | Zona        | Departamemto  | Samples |
#   |--           |--             |--   |
#   | Hot land    | Ica           | 121 |
#   | 		        | Lima          | 110 |
#   |		          | Subtotal      | 231 |
#   | Cold land   | Cajamarca     | 147 |
#   |             | Cuzco         | 89  |
#   |             | Huanuco       | 77  |
#   |             | Junín         | 110 |
#   |	            | Subtotal	    | 423 |
#   | Frozen land | Apurímac      | 121 |
#   |             | Cusco         | 55  |
#   |             | Huancavelica  | 66  |
#   |             | Junín         | 66  |
#   |             | Puno          | 44  |
#   |	            | Subotal       | 333 |
#   |	            | Total         | 987 |
plot(sort(metadata$`Altitude.(masl)`))
abline(h =c(1000, 2000, 3400), col = "red")


#---- lat-lon

min(as.numeric(metadata$`Altitude.(masl)`))
max(as.numeric(metadata$`Altitude.(masl)`))
min(as.numeric(metadata$Longitude))
min(as.numeric(metadata$Latitude))
max(as.numeric(metadata$Latitude))
min(as.numeric(metadata$Longitude))
max(as.numeric(metadata$Longitude))

#---- District 
table(metadata$altzones)

#####----- Elevation Map: Not Working
{
  library(raster)
# ----- get elevation data -------------------------
dem1<- getData("SRTM",lat = 0, lon = -90)
dem2<- getData("SRTM",lat = -20, lon = -65)
dem <- merge(dem1,dem2)

# ----- Country border -----------------------------
gadm <- getData('GADM', country = "PERU", level = 0)
# level = 0 country borders only 
# level = 1 country and department borders
# use st_read() if you have a shapefile

#------ Add cities of reference  -------------------
my_points = data.frame(
  city = c('Lima', 'Cuzco'),
  lon =  c(-77.05,  -72),
  lat =  c(-12.05, -13.5)
)

#------  more libraries  -------------------
library(sf)
my_points = st_as_sf(my_points, coords = c("lon", "lat"), crs = 4326)

# simplified your code a little, using extent() function ############
#create polygon to crop the elevation data file   
Ps1 = as(extent(-80, -69, -18, 0), 'SpatialPolygons')
crs(Ps1) = "+proj=longlat +datum=WGS84 +no_defs"

#crop the elevation data using the polygon
# dem = crop(dem, Ps1, snap= 'out')

#lower the reolution to enable faster plotting
# dem_lower_res <- aggregate(dem, fact=10)
dem.p  <-  rasterToPoints(dem)
df <-  data.frame(dem.p)
colnames(df) = c("lon", "lat", "alt")

#plot the data
library(ggplot2)
ggplot(df) +
  geom_raster(aes(lon, lat, fill = alt))+
  scale_fill_gradientn(colours = terrain.colors(10))+
  geom_sf(data= my_points, color= 'red')+                                         # point
  geom_sf_label(data= my_points, aes(label= city), nudge_y = 0.1,  color= 'gray')+ # label
  geom_tile(data=gadm, aes(long, lat))+
  geom_tile(data= gadm, aes(long, lat), color= 'blue', fill= NA) +             # polygon
  coord_sf(xlim = c(-83, -69), ylim = c(-18, 0))+
  theme_bw()
}
#####
#---- Peruvian Potato Virome

papa1 <- read.xlsx("data/CIP/3_Potato_viroma_all_libaries_completed_Results_mod5.xlsx", sheet=1, startRow = 3)
papa1[1:5,1:15]
papa2 <- read.xlsx("data/CIP/3_Potato_viroma_all_libaries_completed_Results_mod5.xlsx", sheet=2)
papa2[1:5,1:15]
ncol(papa1) == ncol(papa2)
papa <- rbind(papa1, papa2)
papa[1:5,1:15]
dim(papa)
dat=papa

#------
# Plot of mean contig length after removal of 50 nt
plot(dat$Length, (dat$`Coverage_%`),  col="black",
     xlab = "contig length", ylab = "log(mean cov)")
title("Normalized mean \n cov and length\n
      (thr = all)")

#------
alts <- list(nhl <- papa[papa$Sample_code %in% hl$CIP_Code,],
             ncl <- papa[papa$Sample_code %in% cl$CIP_Code,],
             nfl <- papa[papa$Sample_code %in% fl$CIP_Code,])
#------
# for (i in seq_along(alts)){
i=1
dat=alts[[i]]
data <- dat %>%
  select(Library, Place, field, Sample_code, Reference, Length, `Coverage_%`, No.contig, Depth, Depth_norm, 
         Taxonomic_genome_length, `Genome_coverage_%`, New_Genus, New_virus, New_Acronym)
datb <- ddply(data, .(Sample_code, New_Acronym), summarise, cov=mean(Depth))
datc <- na.omit(datb)
datm <- tidyr::spread(datc, Sample_code, cov,  drop=TRUE , fill = 0)
row.names(datm) = datm$New_Acronym
datm= datm[-1]
dim(datm)

#--- Normalizationi
print("normalizing dataset")
dat.mat <- datm/colSums(datm)*100
dat.mat = round(dat.mat, digits = 0)
#--------------------------------------------------------------------------------------
# Calculating metrics ## UNCOMMENT ALL THIS FOR BIPARTITE METRICS
# Bipartite analysis
nsp  <- specieslevel(dat.mat)
#------------------------------------------------------------------------------------------------------------------------------
g = graph.incidence(dat.mat, weighted=T)
V(g)$type
V(g)$name <- c(V(g)$name[1:length(V(g)$type[V(g)$type == "FALSE"])], rep("", length(V(g)$type[V(g)$type == "TRUE"])))
V(g)$size <- c(log(nsp$`lower level`$species.strength+5)*2, log(nsp$`higher level`$degree*5))
V(g)$color <-  c(rep("orange", length(V(g)$type[V(g)$type == "FALSE"])), rep("blue", length(V(g)$type[V(g)$type == "TRUE"])))
E(g)$weight <- 2
shapes = c(rep("circle", length(V(g)$type[V(g)$type == "FALSE"])), rep("square", length(V(g)$type[V(g)$type == "TRUE"])))
p#df(paste0("Network_by_alt_", i,".pdf"), width = 20, height = 20)
plot(g,  vertex.shape=shapes, vertex.size=V(g)$size, vertex.label.cex = 1, vertex.label.color='black', vertex.frame.color="gray", #vertex.label=NA,
     vertex.frame.color="gold",   edge.curved=F,  layout=layout_with_kk(g))
#dev.off()
#}


#-----------------------------------------------------------------------------------------------------------------
# # Identify isolated nodes
# g.cond <- V(g)[igraph::degree(g) == 1]
# gi <- igraph::delete.vertices(g, g.cond) # Remove isolated nodes
# plot(gi, edge.arrow.size=1, vertex.shape=shapes, vertex.label.cex=1.3, vertex.label.color='black', vertex.frame.color="gray", 
#      vertex.frame.color="gold",   edge.curved=F,  layout=layout_with_dh(gi, maxiter=10 ,fineiter = 10, weight.edge.crossings = 1 - sqrt(edge_density(gi))))
# legend(x=-1.3, y=1.1, legend[,2], pch=21,
#        col=legend[,1], pt.bg=legend[,1], pt.cex=1, cex=.8, bty="n", ncol=1)
#------------------------------------------------------------------------------------------------------------------------------
# Bipartite projection 
# Counting the number of interactions
# Virus
g2 <- bipartite.projection(g)
# Vertex names and attributes
bnodes <- as.list(V(g2$proj1))
#V(g2$proj1)$color <- rcPal(14)[as.numeric(cut(as.numeric(factor(as.character(bnodes))), breaks = 14))]
# Edges names and attributes
#E(g2$proj1)$color <- rgPal(10)[as.numeric(cut(E(g2$proj1)$weight, breaks = 10))]
gp1 <- get.adjacency(g2$proj1,sparse=FALSE,attr="weight")
hist()
plot(g2$proj1,edge.width=E(g2$proj1)$weight, vertex.label=V(g2$proj1)$name)
gp2 <- get.adjacency(g2$proj2, sparse=FALSE, attr="weight")
plot(g2$proj2,edge.width=E(g2$proj2)$weight^2, vertex.label=V(g2$proj2)$name)
#--------------------------------------------------------



