# Example of bipartite networks using the data generated by
# Sweet potato virome in Africa. 
# --------------------------------------------------------------------------------------
# Libraries required biparrite and igraph
# If not installed type install.library("bipartite") and install.library("igraph")
# setwd("/Users/ricardoialcala/Documents/Documents-ehecatl/git_db/BiNeVi/iBiNeVI/")
#setwd("/Users/ricardoialcala/Documents/Documents-ehecatl/git_db/BiNeVi_deprecated/iBiNeVI/")
#setwd("/Users/ricardoi//Documents/Documents-ehecatl/git_db/BiNeVi_deprecated/iBiNeVI/")
#@ Path Macbook pro
setwd("/Users/ricardoi/Library/Mobile Documents/com~apple~CloudDocs/Documents-cloud/Documents-ehecatl/git_db/BiNeVi_deprecated/iBiNeVI/")

library(bipartite)
library(igraph)
library(dplyr)
library(plyr)
library(viridis)
library(reshape2)
library(tidyr)
library(openxlsx)
# -----------------------------

# papa <- read.csv("data/CIP/3_Potato_viroma_all_libaries_completed_Results_mod3.csv", header = T, skip=2)
# papa[1:5,1:15]
papa1 <- read.xlsx("data/CIP/3_Potato_viroma_all_libaries_completed_Results_mod5.xlsx", sheet=1, startRow = 3)
papa1[1:5,1:15]
papa2 <- read.xlsx("data/CIP/3_Potato_viroma_all_libaries_completed_Results_mod5.xlsx", sheet=2)
papa2[1:5,1:15]
ncol(papa1) == ncol(papa2)
papa <- rbind(papa1, papa2)
papa[1:5,1:15]
dim(papa)
# 
# # Adding metadata
metadata <- read.xlsx("data/CIP/2_Viroma_web_180319_final.xlsx", sheet = 2)
metadata2 <- metadata %>% select_if(~!all(is.na(.)))
metadata2

table(metadata$Department)

# Manipulating papa data
# First check the structure
str(papa)
#Remove blanks and < 49 
papa$`Genome_coverage_%` <- as.numeric(papa$`Genome_coverage_%`)
# papa$n.Genus <- 
papa$n.Acronym <- paste(papa$New_Acronym, papa$Segment)

# write.csv(papa, "Peru_papa_virome_finaldata20190402.csv")
# After manipulating papa structure to start filtering, I will change the name to dat
# Manipulating dat

for (i in seq(10, 90, 5)){
  print(i)
dat <- papa[papa$`Genome_coverage_%` >= i,]
# 
dat2 <- dat %>%
  select(Library, Place, field, Sample_code, Reference, Length, `Coverage_%`, No.contig, Depth, Depth_norm, 
        Taxonomic_genome_length, `Genome_coverage_%`, New_Genus, New_virus, New_Acronym, n.Acronym)

# dat.att <- ddply(dat, .(New_Acronym, Sample.code), summarise, cov=mean(Coverage))
datb <- ddply(dat2, .(Place, n.Acronym), summarise, cov=mean(Depth))
datc <- na.omit(datb)
datm <- tidyr::spread(datc, Place, cov,  drop=TRUE , fill = 0)
row.names(datm) = datm$n.Acronym
papa.m= datm[-1]
#--------------------------------------------------------------------------------------
# Calculating metrics ## UNCOMMENT ALL THIS FOR BIPARTITE METRICS
# Bipartite analysis
papa.m.sp.table  <- specieslevel(papa.m)
#------------------------------------------------------------------------------------------------------------------------------
g = graph.incidence(papa.m, weighted=T)
V(g)$type
V(g)$name
V(g)$size <- c(sqrt(rowSums(papa.m))/4, papa.m.sp.table$`higher level`$species.strength)
V(g)$color <- c(rep("#ffd450", nrow(papa.m)), rainbow(9))
  # rainbow(ncol)[as.numeric(cut(as.numeric(factor(as.character(papa1$Sample.code))), breaks = ncol))]
E(g)$weight <- 2
shapes = c(rep("circle", length(V(g)$type[V(g)$type == "FALSE"])), rep("square", length(V(g)$type[V(g)$type == "TRUE"])))
pdf(paste("Network_by_department_by_percent_", i,".pdf", sep=""), width = 20, height = 20)
plot(g,  vertex.shape=shapes, vertex.size=V(g)$size, vertex.label.cex = 1.7, vertex.label.color='black', vertex.frame.color="gray", #vertex.label=NA,
     vertex.frame.color="gold",   edge.curved=F,  layout=layout_with_kk(g))
dev.off()
}

##------------------------------------------------------------------------------------------------------------------------------
##------------------------------------------------------------------------------------------------------------------------------
##------------------------------------------------------------------------------------------------------------------------------
##------------------------------------------------------------------------------------------------------------------------------
# solo una red

dat <- papa[papa$`Genome_coverage_%` >= 1,]
# 
dat2 <- dat %>%
  select(Library, Place, field, Sample_code, Reference, Length, `Coverage_%`, No.contig, Depth, Depth_norm, 
         Taxonomic_genome_length, `Genome_coverage_%`, New_Genus, New_virus, New_Acronym, n.Acronym)

# dat.att <- ddply(dat, .(New_Acronym, Sample.code), summarise, cov=mean(Coverage))
datb <- ddply(dat2, .(Place, n.Acronym), summarise, cov=mean(Depth))
datc <- na.omit(datb)
datm <- tidyr::spread(datc, Place, cov,  drop=TRUE , fill = 0)
row.names(datm) = datm$n.Acronym
papa.m= datm[-1]

#--------------------------------------------------------------------------------------
# Calculating metrics ## UNCOMMENT ALL THIS FOR BIPARTITE METRICS
# Bipartite analysis
# Plot bipartite and matrix
plotweb(sortweb(papa.m,sort.order="inc"), method="normal")   
visweb(sortweb(papa.m,sort.order="inc"), type="nested", labsize=2,
        square="interaction", text="none", textsize = 3,circles=F, frame=FALSE)
# # Plot modules
mod.papa <- computeModules(papa.m)
plotModuleWeb(mod.papa)
# Network and node metrics
papa.m.net.table <- bipartite::networklevel(papa.m, weighted = T)
papa.m.sp.table  <- specieslevel(papa.m)
papa.m.nest.table <- nestedcontribution(papa.m)
# papa.m.sp.table$`lower level`$species.strength
# write.csv(papa.m.net.table, "Peru_papa_virome_network_level.csv")
#------------------------------------------------------------------------------------------------------------------------------

g = graph.incidence(papa.m, weighted=T)
V(g)$type
V(g)$name
V(g)$size <- c(sqrt(rowSums(papa.m))/4, papa.m.sp.table$`higher level`$species.strength)
V(g)$color <- c(rep("#ffd450", nrow(papa.m)), rainbow(9))
# rainbow(ncol)[as.numeric(cut(as.numeric(factor(as.character(papa1$Sample.code))), breaks = ncol))]
E(g)$weight <- 2
shapes = c(rep("circle", length(V(g)$type[V(g)$type == "FALSE"])), rep("square", length(V(g)$type[V(g)$type == "TRUE"])))
pdf(paste("Network_by_department_by_percent_", i,".pdf", sep=""), width = 20, height = 20)
plot(g,  vertex.shape=shapes, vertex.size=V(g)$size, vertex.label.cex = 1.7, vertex.label.color='black', vertex.frame.color="gray", #vertex.label=NA,
     vertex.frame.color="gold",   edge.curved=F,  layout=layout_with_kk(g))

 # degreedistr(swp.m)
# #------
# m1 <- LPA_wb_plus(swp.m)
# m1 <- DIRT_LPA_wb_plus(swp.m) # better for weighted networks
# Mod1swp <- convert2moduleWeb(log(swp.m+1), m1)
# plotModuleWeb(Mod1swp, weighted = T)
#--------------------------------------------------------------------------------------
papa.table1 <- papa.m.sp.table$`higher level` %>% 
                select(degree, species.strength, nestedrank, species.specificity.index)
Samples <- as.data.frame(table(metadata$Department))
papa.table1 <- cbind(Samples, papa.table1)
papa.table1



